### DISCUSSION TITLE
All colors on new board are inverted.

### DISCUSSION THREAD

**Discussion by polarkernel** [rzeldent/esp32-smartdisplay#191] [Q&A]
I am very happy to work with _esp32-smartdisplay_. I had great results with the ESP32-JC8048W550C board from Guition (using esp32-8048S050C.json from your site [platformio-espressif32-sunton](https://github.com/rzeldent/platformio-espressif32-sunton)). I used to use _TFT_eSPI_, but since this is not supported by EEZ Studio, I switched to _esp32-smartdisplay_. I also like to have one platform for all my boards.

Now I bought a smaller board that is not yet in this database: ESP32-JC2432W328C, also made by Guition, see [AliExpress](https://de.aliexpress.com/item/1005006732002132.html). It works with ST7789 for TFT and CST816S for touch. Adapting an existing platform file from your database was quite easy, once the pins were known from the Chinese documentation, as usual. Now, it works, well, almost. Only the colors are all inverted!

There are not many color-related definitions in the configuration. I played with the known attributes, like COLOR_SPACE_RGB <-> COLOR_SPACE_BGR and LV_COLOR_16_SWAP. But nothing helped. Unfortunately, my knowledge of the hardware of these devices is very limited.

In _TFT_eSPI_, the solution was simple. You either had to `#define TFT_INVERSION_ON` in the "User_Setup.h" file, or to call `tft.invertDisplay(true);` in the setup() section of the sketch.

My question is now: Does _esp32-smartdisplay_ also provide such a statement? Or does anyone have an idea where the colors could be inverted? I am grateful for any help. It would be very inconvenient to invert all colors in EEZ Studio.

**Comment by rzeldent** [comment]
Yes, color inversion is normally caused by IPSdisplays. Try adding the setting

```"'-D DISPLAY_IPS'"```

If you have a new (working) definition I can add it to the boards!

**Reply by polarkernel** [reply]
> Yes, color inversion is normally caused by IPSdisplays. Try adding the setting
`"'-D DISPLAY_IPS'"`

Thanks for your answer, but it didn't help, the colors are still inverted.

> If you have a new (working) definition I can add it to the boards!

Sure. I already had that in mind. I will also have to solve the touch part, it seems to work only partially.

**Reply by polarkernel** [reply]
It is weird. I tried to observe what happens at the place in the file _lvgl_panel_st7789_spi.c_, where `DISPLAY_IPS` is used. I "instrumented" this code with some additional debug output like this (starting at line 79):

```C
	log_d("\nDISPLAY_IPS is set to %d",DISPLAY_IPS);
    esp_lcd_panel_handle_t panel_handle;
    ESP_ERROR_CHECK(esp_lcd_new_panel_st7789(io_handle, &panel_dev_config, &panel_handle));
	log_d("\n got panel");
    ESP_ERROR_CHECK(esp_lcd_panel_reset(panel_handle));
	log_d("\n reset panel");
    ESP_ERROR_CHECK(esp_lcd_panel_init(panel_handle));
	log_d("\n init panel");
#ifdef DISPLAY_IPS
    // If LCD is IPS invert the colors
	log_d("\nbefore call to invert_color");
    ESP_ERROR_CHECK(esp_lcd_panel_invert_color(panel_handle, true));
	log_d("\after call to invert_color");
#endif
```
But it seems that the execution stops after the creation of `panel_handle`:
```
[     7][D][esp32-hal-cpu.c:244] setCpuFrequencyMhz(): PLL: 480 / 2 = 240 Mhz, APB: 80000000 Hz
[    27][D][esp32_smartdisplay.c:173] smartdisplay_init(): smartdisplay_init  
[    35][D][esp32_smartdisplay.c:93] smartdisplay_led_set_rgb(): R:0, G:0, B:0
[    44][D][lvgl_panel_st7789_spi.c:47] lvgl_lcd_init(): spi_bus_config: mosi_io_num:13, miso_io_num:12, sclk_io_num:14, quadwp_io_num:-1, quadhd_io_num:-1, max_transfer_sz:38400, flags:0x00000000, intr_flags:0x0000
[    64][D][lvgl_panel_st7789_spi.c:66] lvgl_lcd_init(): io_spi_config: cs_gpio_num:15, dc_gpio_num:2, spi_mode:3, pclk_hz:24000000, trans_queue_depth:10, user_ctx:0x3ffcf69c, on_color_trans_done:0x400f7f28, lcd_cmd_bits:8, lcd_param_bits:8, flags:{dc_as_cmd_phase:0, dc_low_on_data:0, octal_mode:0, lsb_first:0}
[    92][D][lvgl_panel_st7789_spi.c:78] lvgl_lcd_init(): panel_dev_config: reset_gpio_num:1, color_space:0, bits_per_pixel:16, flags:{reset_active_high:0}, vendor_config:0x00000000
[   108][D][lvgl_panel_st7789_spi.c:79] lvgl_lcd_init():
DISPLAY_IPS is set to 1
```
Nothing more. However, the display shows the given widgets, just in inverted colors. I would expect an output from `ESP_ERROR_CHECK` in case of an error, but there is nothing.

**Reply by polarkernel** [reply]
Two steps forward to the solution. 

First, the initialization of the panel stopped due to a wrong `ST7789_DEV_CONFIG_RESET_GPIO_NUM` value. I have set it to 1 instead of -1, it seems I was too fast. Now, the initialization completes.

`DISPLAY_IPS` is already defined somewhere! If I omitted it in the board definition, `esp_lcd_panel_invert_color(panel_handle, true)` is called and the display is inverted. If I comment out this line in _lvgl_panel_st7789_spi.c_, the colors are correct. When I set `"'-D DISPLAY_IPS=0'"` in the board definition, I get tons of warnings of the form:
```css
<command-line>: warning: "DISPLAY_IPS" redefined
<command-line>: note: this is the location of the previous definition
```
But there is no definition of `DISPLAY_IPS`, nowhere. A text search over the whole project only returns the lines with `#ifdef DISPLAY_IPS` in the panel files. I have no idea, where this definition comes from. 

So I tried to undefine `DISPLAY_IPS` in the board definition by
```
"'-U DISPLAY_IPS'"
```
and this works! Even if spooky, the problem can be solved via the board definition.

**Comment by polarkernel** [comment]
@rzeldent Now I have working definitions of two new boards, along with pictures of them. How would you like to get them?

**Comment by rzeldent** [comment]
Apparently this was not an IPS board but was defined, so indeed, undefining solved this!

Are you able to make a branch and a pull request? Then I can merge it.... That would be the easiest solution for me...
Nice to see we're adding new boards!

**Comment by polarkernel** [comment]
> Are you able to make a branch and a pull request?

Sorry, but I am very new to GitHub. It would take me days to figure out how to do that. Would a zip file on Dropbox work for you? There is no security issue and I can delete the file and the link once you got it. Are you aware of the fact, that these are Guition boards, and not Sunton?

> Nice to see we're adding new boards!

Well, thank you very much for the great work you have done! I really like your approach, it is very convenient for the user.
